astropy/tests/helper.py

```diff
@@ -412,11 +413,16 @@ def generic_recursive_equality_test(a, b, class_history):
     Check if the attributes of a and b are equal. Then,
     check if the attributes of the attributes are equal.
     """
-    dict_a = a.__getstate__() if hasattr(a, '__getstate__') else a.__dict__
+    if PYTHON_LT_3_11:
+        dict_a = a.__getstate__() if hasattr(a, '__getstate__') else a.__dict__
+    else:
+        # NOTE: The call may need to be adapted if other objects implementing a __getstate__
+        # with required argument(s) are passed to this function.
+        # For a class that has no instance __dict__ and no __slots__, the default state is None
+        dict_a = a.__getstate__(a) if inspect.isclass(a) else a.__getstate__() or dict()
```

Comment:
@dhomeier , do we also need this in `main` and v5.1.x ?

Reply:
Yes, we should – I think in principle any function or other thing that returns a state of `None` might be pickled.

Reply:
Reading about the Python 3.11 [`object.__getstate__()`](https://docs.python.org/3.11/library/pickle.html#object.__getstate__) I am getting the impression that even the patched code here might not behave as expected in the case where a class uses `__slots__`. Should we patch this code even further?

Reply:
Do you know an example of something that implements `__slots__`?
> For a class that has [__slots__](https://docs.python.org/3.11/reference/datamodel.html#object.__slots__) and no instance [__dict__](https://docs.python.org/3.11/library/stdtypes.html#object.__dict__), the default state is a tuple whose first item is None and whose second item is a dictionary mapping slot names to slot values described in the previous bullet.

It _looks_ like we could use `a.__getstate__()[1]` in that case...

Reply:
There are classes in `astropy` that use `__slots__`:
```shell
$ git grep __slots__
CHANGES.rst:4077:  because the class now uses ``__slots__`` any attributes on the class have to
astropy/cosmology/funcs/comparison.py:46:    __slots__ = ("wrapped", )
astropy/table/bst.py:67:    __slots__ = ('val',)
astropy/table/bst.py:108:    __slots__ = ('key', 'data', 'left', 'right')
astropy/table/soco.py:17:    __slots__ = ('key', 'value')
astropy/utils/data_info.py:237:        dct.setdefault('__slots__', [])
astropy/utils/data_info.py:288:    __slots__ = ['_parent_cls', '_parent_ref', '_attrs']
astropy/utils/data_info.py:534:    __slots__ = ['_format_funcs', '_copy_indices']
```

Reply:
And hopefully more will use slots in py3.10+ with ``dataclass(slots=True)``!

Reply:
All those cases also have `__dict__`, so that takes precedence. However there is a weird exception `astropy.utils.data_info`:
```python
from astropy.utils.data_info import DataInfo
DataInfo.__dict__.keys()
dict_keys(['__module__', '__doc__', '_stats', 'attrs_from_parent', 'attr_names', '_attr_defaults', '_attrs_no_copy', '_info_summary_attrs', '__slots__', '_represent_as_dict_attrs', '_construct_from_dict_args', '_represent_as_dict_primary_data', '__init__', '_parent', '__get__', '__set__', '__getstate__', '__setstate__', '_represent_as_dict', '_construct_from_dict', 'info_summary_attributes', 'info_summary_stats', '__call__', '__repr__', '_attrs', '_parent_cls', '_parent_ref', 'description', 'dtype', 'name', 'meta', 'format', 'unit'])

DataInfo.__getstate__(DataInfo)
<member '_attrs' of 'DataInfo' objects>
```
That `member_descriptor` has neither keys nor is it iterable, no idea what use to make of that...

Reply:
Could force it somehow to fall back to `a.__dict__` of course